py/desiutil/redirect.py

```diff
@@ -0,0 +1,247 @@
+
+        # This closes the python file handles, and marks the POSIX
+        # file descriptors for garbage collection- UNLESS those
+        # are the special file descriptors for stderr/stdout.
+        sys.stdout.close()
+        sys.stderr.close()
+
+        # Close fd_out/fd_err if they are open, and copy the
+        # input file descriptors to these.
+        os.dup2(out_to, fd_out)
+        os.dup2(err_to, fd_err)
+
+        # Create a new sys.stdout / sys.stderr that points to the
+        # redirected POSIX file descriptors.  In Python 3, these
+        # are actually higher level IO objects.
+        if sys.version_info[0] < 3:
+            sys.stdout = os.fdopen(fd_out, "wb")
+            sys.stderr = os.fdopen(fd_err, "wb")
+        else:
+            # Python 3 case
+            sys.stdout = io.TextIOWrapper(os.fdopen(fd_out, 'wb'))
+            sys.stderr = io.TextIOWrapper(os.fdopen(fd_err, 'wb'))
+
+        # update DESI logging to use new stdout
+        for name, logger in desi_loggers.items():
+            hformat = None
+            while len(logger.handlers) > 0:
+                h = logger.handlers[0]
+                if hformat is None:
+                    hformat = h.formatter._fmt
+                logger.removeHandler(h)
+            # Add the current stdout.
+            ch = logging.StreamHandler(sys.stdout)
+            formatter = logging.Formatter(hformat, datefmt='%Y-%m-%dT%H:%M:%S')
+            ch.setFormatter(formatter)
+            logger.addHandler(ch)
+
+    def _open_redirect(filename):
+        # Open python file, which creates low-level POSIX file
+        # descriptor.
+        file_handle = open(filename, "w")
+
+        # Redirect stdout/stderr to this new file descriptor.
+        _redirect(out_to=file_handle.fileno(), err_to=file_handle.fileno())
+        return file_handle
+
+    def _close_redirect(handle):
+        # Close python file handle, which will mark POSIX file descriptor for
+        # garbage collection.  That is fine since we are about to overwrite those.
+        if handle is not None:
+            handle.close()
+
+        # Flush python handles for good measure
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        try:
+            # Restore old stdout and stderr
+            _redirect(out_to=saved_fd_out, err_to=saved_fd_err)
+        except:
+            pass
+
+    # Redirect both stdout and stderr to the same file
+
+    if to is None:
+        to = "/dev/null"
+
+    if rank == 0:
+        log = get_logger()
+        log.info("Begin log redirection to {} at {}".format(to, time.asctime()))
+
+    # Try to open the redirected file.
+
+    pto = to
+    if to != "/dev/null" and comm is not None:
+        pto = "{}_{}".format(to, rank)
+
+    fail_open = 0
+    file = None
+    try:
+        file = _open_redirect(pto)
+    except:
+        log = get_logger()
+        log.error(
+            "Failed to open redirection file {} at {}".format(pto, time.asctime())
```

Comment:
The logging framework actually prefers `%` based formatting, *e.g.*:
```
python
log.error("Failed to open redirection file %s at %s.", pto, time.asctime())
```
This style is used through most of the rest of desiutil.  It also allows the logging framework to perform certain optimizations.

Reply:
Done